### PR TITLE
[fix] [ml] Topics stats shows msgBacklog but there reality no backlog

### PR DESCRIPTION
**Pull request by poorbarcode**
#### 1. `readPosition` point to a deleted ledger

When `trim ledgers` and `create new cursor` are executed concurrently, it will cause the `readPosition` of the cursor to point to a deleted ledger.

| time | `trim ledgers` | `create new cursor` |
| --- | --- | --- |
| 1 | | set read position and mark deleted position |
| 2 | delete ledger | |
| 3 | | add the cursor to `ManagedLedger.cursors` |

----

#### 2. Backlog wrong caused by `readPosition` wrong
<strong>(Highlight)</strong>Since the read position of the cursor is pointing at a deleted ledger, so deleted messages will never be consumed or acknowledged. Since the backlog in the API `topics stats` response is calculated as this: `managedLedger.entriesAddedCounter - cursor.messagesConsumedCounter`, the result is: Topics stats show `msgBacklog` but there is reality no backlog.
- `managedLedger.entriesAddedCounter`: Pulsar will set it to `0` when creating a new managed ledger, it will increment when adding entries.
- `cursor.messagesConsumedCounter`: Pulsar will set it to `0` when creating a new cursor, it will increment when acknowledging.

For example: 
- write entries to the managed ledger: `{1:0~1:9}...{5:0~5:9}`
  - `managedLedger.entriesAddedCounter` is `50` now
- create a new cursor, and set the read position to `1:0`
  -  `cursor.messagesConsumedCounter` is `0` now
- delete ledgers `1~4`
- consume all messages
  - can only consume the messages {5:0~5:9}, so `cursor.messagesConsumedCounter` is `10` now
- the `backlog` in response of `topics stats` is `50 - 10 = 40`, but there reality no backlog

----

#### 3. Reproduce issue
Sorry, I spent 4 hours trying to write a non-invasive test, but failed. <strong>(Highlight)</strong>You can reproduce by `testBacklogIfCursorCreateConcurrentWithTrimLedger` in the PR #19274

https://github.com/apache/pulsar/blob/a2cdc759fc2710e4dd913eb0485d23ebcaa076a4/pulsar-broker/src/test/java/org/apache/bookkeeper/mledger/impl/StatsBackLogTest.java#L163

### Modifications

Avoid the race condition of `cursor.initializeCursorPosition` and `internalTrimLedgers`

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: 
- https://github.com/poorbarcode/pulsar/pull/53

### CODE REVIEW

**Review by lhotari — APPROVED** [review]

**Comment by HQebupt** [comment]
> consume all messages
> - can only consume the messages {5:0~5:10}, so managedLedger.entriesAddedCounter is 5 now
> the backlog in response of topics stats is 50 - 5 = 45, but there reality no backlog

I believe the expected behavior is following. 
```
can only consume the messages {5:0~5:10}, so managedLedger.entriesAddedCounter is ~~5~~ **10** now
the backlog in response of topics stats is ~~50 - 5 = 45~~ **50 - 10 = 40**, but there reality no backlog
```

**Review by codelipenghui — APPROVED** [review]
@poorbarcode Great finding!

The change looks good to me.
You should update the `Modification` section of the PR description.
Now the solution is to avoid the race condition of `cursor.initializeCursorPosition` and `internalTrimLedgers`, right?

**Comment by poorbarcode** [comment]
@codelipenghui 

> Now the solution is to avoid the race condition of cursor.initializeCursorPosition and internalTrimLedgers, right?

Yes, Already fixed

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -690,11 +690,11 @@ private void recoveredCursor(PositionImpl position, Map<String, Long> properties
         }
         log.info("[{}] Cursor {} recovered to position {}", ledger.getName(), name, position);
         this.cursorProperties = cursorProperties;
-        messagesConsumedCounter = -getNumberOfEntries(Range.openClosed(position, ledger.getLastPosition()));
         markDeletePosition = position;
         persistentMarkDeletePosition = position;
         inProgressMarkDeletePersistPosition = null;
         readPosition = ledger.getNextValidPosition(position);
+        messagesConsumedCounter = -getNumberOfEntries(Range.openClosed(position, ledger.getLastPosition()));
```
Comment (HQebupt) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:697]: Why change the order of this line? it doesn't seem to help.

**Comment by poorbarcode** [comment]
Hi @HQebupt 

>I believe the expected behavior is following : can only consume the messages {5:0~5:10}, so managedLedger.entriesAddedCounter is 5 10 now. the backlog in response of topics stats is 50 - 5 = 45 50 - 10 = 40, but there reality no backlog

I wrote the wrong variable name. It has been corrected. Thanks

**Review by HQebupt — APPROVED** [review]
LGTM

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java (hunk 1 above)
Comment (poorbarcode) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: This line doesn't really have anything to do with this issue, I removed this change. Please take a look

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 2]
```diff
@@ -993,12 +993,11 @@ public synchronized void asyncOpenCursor(final String cursorName, final InitialP
             public void operationComplete() {
                 log.info("[{}] Opened new cursor: {}", name, cursor);
                 cursor.setActive();
```
Comment (mattisonchao) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:995]: Should we move this line to `synchromized`?

**Review by mattisonchao — APPROVED** [review]
LGTM, thanks for your contribution.

**Pull request merged**